lib/pci_caps.c

```diff
@@ -0,0 +1,550 @@
+{
+    vfu_pci_config_space_t *config_space;
+    uint8_t *prevp;
+    size_t offset;
+
+    config_space = vfu_pci_get_config_space(vfu_ctx);
+
+    prevp = &config_space->hdr.cap;
+
+    if (cap->off != 0) {
+        if (cap->off < PCI_STD_HEADER_SIZEOF) {
+            vfu_log(vfu_ctx, LOG_ERR, "invalid offset %#lx for capability "
+                    "%u (%s)\n", cap->off, cap->id, cap->name);
+            return EINVAL;
+        }
+
+        if (cap_find_by_offset(vfu_ctx, cap->off, cap->size) != NULL) {
+            vfu_log(vfu_ctx, LOG_ERR, "overlap found for capability "
+                    "%u (%s)\n", cap->id, cap->name);
+            return EINVAL;
+        }
+
+        while (*prevp != 0) {
+            prevp = pci_config_space_ptr(vfu_ctx, *prevp + PCI_CAP_LIST_NEXT);
+        }
+    } else if (*prevp == 0) {
+        cap->off = PCI_STD_HEADER_SIZEOF;
+    } else {
+        for (offset = *prevp; offset != 0; offset = *prevp) {
+            uint8_t id;
+            size_t size;
+
+            id = *pci_config_space_ptr(vfu_ctx, offset + PCI_CAP_LIST_ID);
+            prevp = pci_config_space_ptr(vfu_ctx, offset + PCI_CAP_LIST_NEXT);
+
+            if (*prevp == 0) {
+                size = cap_size(id, pci_config_space_ptr(vfu_ctx, offset));
+                cap->off = ROUND_UP(offset + size, 4);
+                break;
+            }
+        }
+    }
+
+    if (cap->off + cap->size >
+        vfu_ctx->reg_info[VFU_PCI_DEV_CFG_REGION_IDX].size) {
+        vfu_log(vfu_ctx, LOG_ERR, "no config space left for capability "
+                "%u (%s) of size %zu bytes at offset %#lx\n", cap->id,
+                cap->name, cap->size, cap->off);
+        return ENOSPC;
+    }
+
+    memcpy(cap_data(vfu_ctx, cap), data, cap->size);
+    /* Make sure the previous cap's PCI_CAP_LIST_NEXT points to us. */
+    *prevp = cap->off;
+    /* Make sure our PCI_CAP_LIST_NEXT is zeroed. */
+    *pci_config_space_ptr(vfu_ctx, cap->off + PCI_CAP_LIST_NEXT) = 0;
+    return 0;
+}
+
+ssize_t
+vfu_pci_add_capability(vfu_ctx_t *vfu_ctx, size_t pos, int flags, void *data)
+{
+    size_t space_size = vfu_ctx->reg_info[VFU_PCI_DEV_CFG_REGION_IDX].size;
+    struct pci_cap cap;
+    int ret;
+
+    assert(vfu_ctx != NULL);
+
+    if (flags & ~(VFU_CAP_FLAG_EXTENDED | VFU_CAP_FLAG_CALLBACK |
+        VFU_CAP_FLAG_READONLY)) {
+        return ERROR(EINVAL);
+    }
+
+    if ((flags & VFU_CAP_FLAG_CALLBACK) && (flags & VFU_CAP_FLAG_READONLY)) {
+        return ERROR(EINVAL);
+    }
+
+    if ((flags & VFU_CAP_FLAG_CALLBACK) &&
+        vfu_ctx->reg_info[VFU_PCI_DEV_CFG_REGION_IDX].cb == NULL) {
+        return ERROR(EINVAL);
+    }
+
+    if ((flags & VFU_CAP_FLAG_EXTENDED)) {
+        return ERROR(ENOTSUP);
+    }
+
+    if (vfu_ctx->pci.nr_caps == VFU_MAX_CAPS) {
+        return ERROR(ENOSPC);
+    }
+
+    cap.id = ((struct cap_hdr *)data)->id;
+    cap.hdr_size = sizeof (struct cap_hdr);
+    cap.size = cap_size(cap.id, data);
```

Comment:
Maybe move this after id validation below. So that we can assert() for un-supported id in cap_size().

Reply:
good idea